src/toxcore/dht_new/dht_impl.rs

```diff
@@ -0,0 +1,242 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Implements Dht related structures.
+*/
+
+use std::collections::HashMap;
+use toxcore::dht_new::packet::*;
+use toxcore::dht_new::packed_node::*;
+use toxcore::dht_new::kbucket::*;
+use toxcore::dht_new::binary_io::*;
+use toxcore::dht_new::codec::*;
+use toxcore::crypto_core::*;
+use toxcore::dht_new::packet_kind::*;
+use std::io::Error;
+use nom::IResult;
+use bincode::{serialize, deserialize, Infinite};
+
+/// Manage hash table for precomputed keys.
+#[derive(Clone, Eq, Debug, PartialEq)]
+pub struct PrecomputedKeys {
+    cache: HashMap<Vec<u8>, Vec<u8>>,
+}
+
+impl PrecomputedKeys {
+    /// manage hash table for precomputed keys
+    pub fn new () -> PrecomputedKeys {
+        PrecomputedKeys {
+            cache: HashMap::new(),
+        }
+    }
+
+    /// Get precomputed keys
+    /// If the Key is not found in cache, create symmetric key and insert it into cache for later use.
+    pub fn get_symmetric_key (&mut self, key_pair: (&SecretKey, &PublicKey)) -> Option<PrecomputedKey> {
```

Comment:
Why Option?

Reply:
I will change return type from Option to Result to handle error better.